avogadro/core/cube.cpp

```diff
@@ -194,6 +206,165 @@ float Cube::value(int i, int j, int k) const
+
+        // std::cout << m_spacing[0];
+    if (i == 0)
+    {
+        x[0][0] = m_data[dataIdx + 1];
+        x[0][1] = m_data[dataIdx];
+        run[0] = m_spacing.x();
+    }
+    else if (i == (m_points.x() - 1))
+    {
+        x[0][0] = m_data[dataIdx];
+        x[0][1] = m_data[dataIdx - 1];
+        run[0] = m_spacing.x();
+    }
+    else
+    {
+        x[0][0] = m_data[dataIdx + 1];
+        x[0][1] = m_data[dataIdx - 1];
+        run[0] = 2 * m_spacing.x();
+    }
+
+    if (j == 0)
+    {
+        x[1][0] = m_data[dataIdx + m_points.x()];
+        x[1][1] = m_data[dataIdx];
+        run[1] = m_spacing.y();
+    }
+    else if (j == (m_points.y() - 1))
+    {
+        x[1][0] = m_data[dataIdx];
+        x[1][1] = m_data[dataIdx - m_points.x()];
+        run[1] = m_spacing.y();
+    }
+    else
+    {
+        x[1][0] = m_data[dataIdx + m_points.x()];
+        x[1][1] = m_data[dataIdx - m_points.x()];
+        run[1] = 2 * m_spacing.y();
+    }
+
+    if (k == 0)
+    {
+        x[2][0] = m_data[dataIdx + (m_points.x() * m_points.y())];
+        x[2][1] = m_data[dataIdx];
+        run[2] = m_spacing.z();
+    }
+    else if (k == (m_points.z() - 1))
+    {
+        x[2][0] = m_data[dataIdx];
+        x[2][1] = m_data[dataIdx - (m_points.x() * m_points.y())];
+        run[2] = m_spacing.z();
+    }
+    else
+    {
+        x[2][0] = m_data[dataIdx + (m_points.x() * m_points.y())];
+        x[2][1] = m_data[dataIdx - (m_points.x() * m_points.y())];
+        run[2] = 2 * m_spacing.z();
+    }
+
+    std::array<float, 3> ret;
+
+    ret[0] = (x[0][1] - x[0][0]) / run[0];
+    ret[1] = (x[1][1] - x[1][0]) / run[1];
+    ret[2] = (x[2][1] - x[2][0]) / run[2];
+
+    return ret;
+}
+
+std::array<std::array<float, 3>, 8>
+Cube::getGradCube(int i, int j, int k) const
+{
+    std::array<std::array<float, 3>, 8> grad;
+
+    grad[0] = computeGradient(i, j, k);
+    grad[1] = computeGradient(i + 1, j, k);
+    grad[2] = computeGradient(i + 1, j + 1, k);
+    grad[3] = computeGradient(i, j + 1, k);
+    grad[4] = computeGradient(i, j, k + 1);
+    grad[5] = computeGradient(i + 1, j, k + 1);
+    grad[6] = computeGradient(i + 1, j + 1, k + 1);
+    grad[7] = computeGradient(i, j + 1, k + 1);
+
+    return grad;
+}
+
+std::array<float, 8> Cube::getValsCube(int i, int j, int k) const
+{
+    std::array<float, 8> vals;
+    
+    vals[0] = getData(i, j, k);
+    vals[1] = getData(i + 1, j, k);
+    vals[2] = getData(i + 1, j + 1, k);
+    vals[3] = getData(i, j + 1, k);
+    vals[4] = getData(i, j, k + 1);
+    vals[5] = getData(i + 1, j, k + 1);
+    vals[6] = getData(i + 1, j + 1, k + 1);
+    vals[7] = getData(i, j + 1, k + 1);
+
+    return vals;
+}
+
+
+inline float
+Cube::getData(int i, int j, int k) const
+{
+    return m_data[(k * m_points.x() * m_points.y()) + (j * m_points.x()) + i];
+}
+
+
+std::array<std::array<float, 3>, 8> Cube::getPosCube(int i, int j, int k) const
```

Comment:
Hmm.... I just overlooked this bug. I think here's where the issue lies. You know, when you just renders a single molecule it's position comes out to be correct, but when you have more than one mesh it's orientation is really weird. I'll check that now..... And will try to fix that.

Reply:
Like, try to render just a CH4 it will give the correct position but as you will have more than one molecule.. It will give you wrong orientation